public/app/plugins/datasource/cloudwatch/ConfigEditor.tsx

```diff
@@ -0,0 +1,389 @@
+      .loadDatasource(this.state.config.name)
+      .then((ds: CloudWatchDatasource) => {
+        return ds.getRegions();
+      })
+      .then(
+        (regions: any) => {
+          this.setState({
+            regions: regions.map((region: any) => {
+              return {
+                value: region.value,
+                label: region.text,
+              };
+            }),
+          });
+        },
+        (err: any) => {
+          console.error('failed to get latest regions');
+        }
+      );
+  }
+
+  updateDatasource = async (config: any) => {
+    for (const j in config.jsonData) {
+      if (config.jsonData[j].length === 0) {
+        delete config.jsonData[j];
+      }
+    }
+
+    for (const m in config.editorJsonData) {
+      if (!config.hasOwnProperty('jsonData')) {
+        config.jsonData = {};
+      }
+      if (config.editorJsonData[m].length === 0) {
+        if (config.hasOwnProperty('jsonData') && config.jsonData.hasOwnProperty(m)) {
+          delete config.jsonData[m];
+        }
+      } else {
+        config.jsonData[m] = config.editorJsonData[m];
+      }
+    }
+
+    for (const k in config.secureJsonData) {
+      if (config.secureJsonData[k].length === 0) {
+        delete config.secureJsonData[k];
+      }
+    }
+
+    for (const l in config.editorSecureJsonData) {
+      if (!config.hasOwnProperty('secureJsonData')) {
+        config.secureJsonData = {};
+      }
+      if (config.editorSecureJsonData[l].length === 0) {
+        if (config.hasOwnProperty('secureJsonData') && config.secureJsonData.hasOwnProperty(l)) {
+          delete config.secureJsonData[l];
+        }
+      } else {
+        config.secureJsonData[l] = config.editorSecureJsonData[l];
+      }
+    }
+
+    this.props.onOptionsChange({
+      ...config,
+    });
+  };
+
+  onAuthProviderChange = (authType: SelectableValue<string>) => {
+    this.updateDatasource({
+      ...this.state.config,
+      editorJsonData: {
+        ...this.state.config.editorJsonData,
+        authType: authType.value,
+      },
+    });
+  };
+
+  onRegionChange = (defaultRegion: SelectableValue<string>) => {
+    this.updateDatasource({
+      ...this.state.config,
+      editorJsonData: {
+        ...this.state.config.editorJsonData,
+        defaultRegion: defaultRegion.value,
+      },
+    });
+  };
+
+  onResetAccessKey = () => {
+    this.updateDatasource({
+      ...this.state.config,
+      version: this.state.config.version + 1,
+      secureJsonFields: {
+        ...this.state.config.secureJsonFields,
+        accessKey: false,
+      },
+    });
+  };
+
+  onAccessKeyChange = (accessKey: string) => {
+    this.updateDatasource({
+      ...this.state.config,
+      editorSecureJsonData: {
+        ...this.state.config.editorSecureJsonData,
+        accessKey,
+      },
+    });
+  };
+
+  onResetSecretKey = () => {
+    this.updateDatasource({
+      ...this.state.config,
+      version: this.state.config.version + 1,
```

Comment:
This version increase looks strange, this should only be done in the backend I think

Reply:
When I did this with Azure Monitor, it seemed to expect that I would send the version up 1 already, else it would send back an err.

Reply:
that is super strange, when I click Save and Test for other data sources they send back the same value they get from backend. Same thing if I test saving Cloudwatch in master it sends the same version it was loaded with (and gets back version + 1 in the response) 

The backend increments this with one on save. So this not sure what the error could be and what is going on. Maybe some double save? Can you check the network traffic, debug this to figure out why this error is happening?

Reply:
I checked on this one and it was fine so I took it out. Thanks!